xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -406,6 +417,49 @@ void cancelEndpointDataWatch(String clusterName, EndpointWatcher watcher) {
     }
   }
 
+  @Override
+  void watchListenerData(int port, ListenerWatcher watcher) {
+    checkState(configWatcher == null,
+        "ListenerWatcher cannot be set when ConfigWatcher set");
+    checkState(listenerWatcher == null, "ListenerWatcher already registered");
+    listenerWatcher = checkNotNull(watcher, "watcher");
+    checkArgument(port > 0, "port needs to be > 0");
+    this.listenerPort = port;
+    logger.log(XdsLogLevel.INFO, "Started watching listener for port {0}", port);
+    if (rpcRetryTimer != null && rpcRetryTimer.isPending()) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (adsStream == null) {
+      startRpcStream();
+    }
+    updateNodeMetadataForListenerRequest(port);
+    adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, getListOfResourceNames(ldsResourceName));
```

Comment:
I am confused about this method. Isn't `ldsResourceName` always `null` when `XdsClient` is used to watch Listener data? Then this should just be an empty list as always. Why do we need this helper method?

Reply:
In this particular case you are correct: I can just use `ImmutableList.of()` or `Collections.emptyList()`. But calls to `sendAckRequest` and `sendNackRequest`  used `ImmutableList.of(resourceName)` which generates NPE when `resourceName == null`. So I used `getListOfResourceNames(ldsResourceName)` everywhere.

Reply:
You can use empty list for `sendAckRequest()` and `sendNackRequest()` as well.

Reply:
On line 1158, there is `RpcRetryTask.run()` doing this:
```
adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName))
```
How do you want to handle that? Of course I can inline `getListOfResourceNames()` everywhere but looks ugly.